### ISSUE TITLE
Export not found during service deployment

### ISSUE THREAD

**Issue by mijolabs** [aws/copilot-cli#4956]
I have started getting the error `No export named my_app-dev-PublicSubnets found. Rollback requested by user.` during service deployment.

I am not using public subnets and just press enter when prompted to select one. I'm puzzled because this exact configuration has worked before with multiple other apps which don't have public subnets configured, either. 

I believe I'm following the exact same steps that I always have performed when initializing and deploying the app, environment, and service. Looking at my other apps' Exports shows they don't have any `-PublicSubnets` export, so I guess the question is not Why the export isn't being created, but Why is that export expected during deployment now?

Have there been any changes lately or could it potentially be caused by any updates that may have been done to my organization's IAM configuration? Suggestions for resolving this issue would be greatly appreciated.

I'm using Copilot CLI 1.28.

**Comment by bvtujo** [comment]
Hey @mijolabs, could you share a little more about your environment configuration and the service manifest that is causing this error? Specifically, the environment manifest would be very helpful, along with the output of `copilot env show --resources --json`.

It looks to me like a  possible bug where we don't always check for public subnet exports in the environment before enabling a feature which uses them, but I'd need to look more closely to confirm.

**Comment by mijolabs** [comment]
Hi @bvtujo 

Please see my configuration and manifests below.

Environment manifest:
```yaml
name: dev
type: Environment

network:
  vpc:
    id: vpc-xxxxxxxxxxxxxxxxx
    subnets:
      private:
        - id: subnet-xxxxxxxxxxxxxxxxx
        - id: subnet-xxxxxxxxxxxxxxxxx

observability:
  container_insights: false
```

Service manifest:
```yaml
name: my_service
type: Worker Service

image:
  build: path/to/Dockerfile

cpu: 256
memory: 512
platform: linux/x86_64
exec: true

count:
  spot: 1
```


Output of `copilot env show --resources --json`:
```json
{
  "environment":
    {
      "app": "my_app",
      "name": "dev",
      "region": "eu-north-1",
      "accountID": "[ACCOUNT-ID]",
      "registryURL": "",
      "executionRoleARN": "arn:aws:iam::[ACCOUNT-ID]:role/my_app-dev-CFNExecutionRole",
      "managerRoleARN": "arn:aws:iam::[ACCOUNT-ID]:role/my_app-dev-EnvManagerRole",
    },
  "services":
    [
      {
        "app": "my_app",
        "name": "my_service",
        "type": "Worker Service",
      },
    ],
  "jobs": null,
  "tags": { "copilot-application": "my_app", "copilot-environment": "dev" },
  "resources":
    [
      {
        "type": "AWS::IAM::Role",
        "physicalID": "my_app-dev-CFNExecutionRole",
        "logicalID": "CloudformationExecutionRole",
      },
      {
        "type": "AWS::ECS::Cluster",
        "physicalID": "my_app-dev-Cluster-Y50SmXvyJQhU",
        "logicalID": "Cluster",
      },
      {
        "type": "AWS::IAM::Role",
        "physicalID": "my_app-dev-EnvManagerRole",
        "logicalID": "EnvironmentManagerRole",
      },
      {
        "type": "AWS::EC2::SecurityGroup",
        "physicalID": "sg-xxxxxxxxxxxxxxxxx",
        "logicalID": "EnvironmentSecurityGroup",
      },
      {
        "type": "AWS::EC2::SecurityGroupIngress",
        "physicalID": "EnvironmentSecurityGroupIngressFromSelf",
        "logicalID": "EnvironmentSecurityGroupIngressFromSelf",
      },
      {
        "type": "AWS::ServiceDiscovery::PrivateDnsNamespace",
        "physicalID": "ns-xxxxxxxxxxxxxxxxx",
        "logicalID": "ServiceDiscoveryNamespace",
      },
    ],
  "environmentVPC":
    {
      "id": "vpc-xxxxxxxxxxxxxxxxx",
      "publicSubnetIDs": null,
      "privateSubnetIDs":
        ["subnet-xxxxxxxxxxxxxxxxx", "subnet-xxxxxxxxxxxxxxxxx"],
    },
}
```

As mentioned, I've been using the same manifest configurations and performed the same deployment steps for multiple apps prior to this. It's been working great until now and as far as I can tell nothing has changed, although there is a possibility that my organization has modified something in AWS IAM. Not sure if that could cause this issue.

I tested this with Copilot CLI version 1.27 first, and then upgraded to 1.28 see if that would help. I also tested with both service type "Backend Service" and "Worker Service", both yielding the same result.

Please let me know if there is any additional information that I can provide to aid in troubleshooting.

**Comment by bvtujo** [comment]
I think I see the problem. Copilot places all services, even "backend" or "worker" services, in public subnets by default. This is so that you don't incur unnecessary NAT charges and can still have internet connectivity if your workload needs it. 

If you want to specifically place your services in your private subnets only, you'll want to use the [`network.vpc.placement` parameter in the manifest](https://aws.github.io/copilot-cli/docs/manifest/backend-service/#network-vpc-placement). 

For example, you can amend your manifest to include the following:

```yaml
network:
  vpc:
    placement: private
```

This will cause Copilot to launch your tasks only in the private subnets of your VPC and should resolve your error. 

The caveat here, though, is that you'll have to set up NAT or VPC endpoints on your own if your workloads need connectivity with AWS or the internet. Since you're importing your VPC and subnets, Copilot can't manage that for you.

**Comment by bvtujo** [comment]
Did you recently upgrade your copilot version? And if so, which version did you upgrade from?

**Comment by mijolabs** [comment]
I added `network.vpc.placement: private` and it solved the issue. Awesome!

I remember that I used to include that configuration but it is no longer in the manifest files of my other services. I suppose that block must have been removed by accident somehow. It seems like this issue was caused by my misconfiguration.

Thanks a bunch for investigating and resolving it.

Regarding your other question, today I upgraded from 1.27 -> 1.28. It is possible that an earlier version was used when deploying my other services earlier this year, but I cannot remember what version I was on prior to 1.27.
